lib/resolve-exports.js

```diff
@@ -0,0 +1,178 @@
+var path = require('path');
+
+function makeError(code, message) {
+    var error = new Error(message);
+    error.code = code;
+    return error;
+}
+
+function validateExports(exports, basePath) {
+    var isConditional = true;
+
+    if (typeof exports === 'object' && !Array.isArray(exports)) {
+        var exportKeys = Object.keys(exports);
+
+        for (var i = 0; i < exportKeys.length; i++) {
+            var isKeyConditional = exportKeys[i][0] !== '.';
+            if (i === 0) {
+                isConditional = isKeyConditional;
+            } else if (isKeyConditional !== isConditional) {
+                throw makeError('ERR_INVALID_PACKAGE_CONFIG', 'Invalid package config ' + basePath + path.sep + 'package.json, '
+                    + '"exports" cannot contain some keys starting with \'.\' and some not. '
+                    + 'The exports object must either be an object of package subpath keys '
+                    + 'or an object of main entry condition name keys only.');
+            }
+        }
+    }
+
+    if (isConditional) {
+        return { '.': exports };
+    } else {
+        return exports;
+    }
+}
+
+function validateEnvNames(names) {
+    // TODO If exports contains any index property keys, as defined in ECMA-262 6.1.7 Array Index, throw an Invalid Package Configuration error.
+    return names;
+}
+
+var startsWith;
+
+if (typeof String.prototype.startsWith === 'function') {
+    startsWith = function (path, exportedPath) {
+        return path.startsWith(exportedPath);
+    };
+} else {
+    startsWith = function (path, exportedPath) {
+        for (var i = 0; i < exportedPath.length; i++) {
+            if (path[i] !== exportedPath[i]) {
+                return false;
+            }
+        }
+
+        return true;
+    };
+}
```

Comment:
we should maybe just use `string.prototype.startswith` rather than writing our own polyfill

Reply:
Can do, but that would require changing what versions of node the tests run on.

Reply:
I meant, use the npm package, which works on every version of node.